dizoo/common/policy/md_ppo.py

```diff
@@ -1,57 +1,113 @@
+                    loss.append(ppo_loss)
+                    info.append(ppo_info)
+                policy_loss = sum([item.policy_loss for item in loss]) / action_num
+                value_loss = sum([item.value_loss for item in loss]) / action_num
+                entropy_loss = sum([item.entropy_loss for item in loss]) / action_num
+            else:
+            # Calculate ppo error
+                ppodata = ppo_data(
+                    output['logit'], data['logit'], data['action'], output['value'], data['value'], adv, return_,
+                    data['weight']
+                )
+                ppo_loss, ppo_info = ppo_error(ppodata, self._clip_ratio)
+            wv, we = self._value_weight, self._entropy_weight
+            total_loss = ppo_loss.policy_loss + wv * ppo_loss.value_loss - we * ppo_loss.entropy_loss
 
-@POLICY_REGISTRY.register('md_ppo')
-class MultiDiscretePPOPolicy(PPOPolicy):
+        else:
+            output = self._learn_model.forward(data['obs'], mode='compute_actor')
+            adv = data['adv']
+            if self._adv_norm:
+                # Normalize advantage in a total train_batch
+                adv = (adv - adv.mean()) / (adv.std() + 1e-8)
+            
+            if isinstance(data['logit'], list):
+                action_num = len(data['logit'])
+                loss, info = [], []
+                for i in range(action_num):
+                    ppodata = ppo_policy_data(output['logit'][i], data['logit'][i], data['action'][i], adv, data['weight'])
+                    ppo_policy_loss, ppo_info = ppo_policy_error(ppodata, self._clip_ratio)
+                    loss.append(ppo_policy_loss)
+                    info.append(ppo_info)
+                ppo_policy_loss = sum([item.policy_loss for item in loss]) / action_num
+                entropy_loss = sum([item.entropy_loss for item in loss]) / action_num
 
-    def _forward_learn(self, data: dict) -> Dict[str, Any]:
-        output = self._armor.forward(data['obs'], param={'mode': 'compute_action_value'})
-        adv = data['adv']
-        if self._use_adv_norm:
-            adv = (adv - adv.mean()) / (adv.std() + 1e-8)
-        return_ = data['value'] + adv
-        action_num = len(data['logit'])
-        loss, info = [], []
-        for i in range(action_num):
-            data_ = ppo_data(
-                output['logit'][i], data['logit'][i], data['action'][i], output['value'], data['value'], adv, return_,
-                data['weight']
+            else:
+            # Calculate ppo error
+                ppodata = ppo_policy_data(output['logit'], data['logit'], data['action'], adv, data['weight'])
+                ppo_policy_loss, ppo_info = ppo_policy_error(ppodata, self._clip_ratio)
+            wv, we = self._value_weight, self._entropy_weight
+            next_obs = data.get('next_obs')
+            value_gamma = data.get('value_gamma')
+            reward = data.get('reward')
```

Comment:
what if we get `None` values here? Will `None` still works in the following pipeline?

Reply:
ppo deleted